File: crates/anchor-wallet-lib/src/wallet.rs
//! ANCHOR Wallet implementation

use bitcoin::{Address, Amount, Network, ScriptBuf, Txid};
use bitcoincore_rpc::{Auth, Client, RpcApi};

use crate::config::WalletConfig;
use crate::error::{Result, WalletError};
use crate::transaction::{AnchorTransaction, TransactionBuilder};
use anchor_core::AnchorKind;

/// UTXO information
#[derive(Debug, Clone)]
pub struct Utxo {
    /// Transaction ID
    pub txid: Txid,
    /// Output index
    pub vout: u32,
    /// Value in satoshis
    pub amount: u64,
    /// Script pubkey
    pub script_pubkey: ScriptBuf,
    /// Number of confirmations
    pub confirmations: u32,
}

/// Wallet balance
#[derive(Debug, Clone, Default)]
pub struct Balance {
    /// Confirmed balance in satoshis
    pub confirmed: u64,
    /// Unconfirmed balance in satoshis
    pub unconfirmed: u64,
    /// Total balance in satoshis
    pub total: u64,
}

/// ANCHOR wallet for creating and broadcasting messages
pub struct AnchorWallet {
    config: WalletConfig,
    client: Client,
}

impl AnchorWallet {
    /// Create a new wallet connected to a Bitcoin Core node
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use anchor_wallet_lib::{AnchorWallet, WalletConfig};
    ///
    /// let config = WalletConfig::regtest("http://127.0.0.1:18443", "user", "pass");
    /// let wallet = AnchorWallet::new(config)?;
    /// ```
    pub fn new(config: WalletConfig) -> Result<Self> {
        config.validate()?;

        let auth = Auth::UserPass(config.rpc_user.clone(), config.rpc_password.clone());
        
        let client = if let Some(ref wallet_name) = config.wallet_name {
            let url = format!("{}/wallet/{}", config.rpc_url, wallet_name);
            Client::new(&url, auth)?
        } else {
            Client::new(&config.rpc_url, auth)?
        };

        Ok(Self { config, client })
    }

    /// Get the wallet configuration
    pub fn config(&self) -> &WalletConfig {
        &self.config
    }

    /// Get the network
    pub fn network(&self) -> Network {
        self.config.network
    }

    /// Get wallet balance
    pub fn get_balance(&self) -> Result<Balance> {
        let balances = self.client.get_balances()?;
        
        let confirmed = balances.mine.trusted.to_sat();
        let unconfirmed = balances.mine.untrusted_pending.to_sat();
        
        Ok(Balance {
            confirmed,
            unconfirmed,
            total: confirmed + unconfirmed,
        })
    }

    /// Get a new receiving address
    pub fn get_new_address(&self) -> Result<Address<bitcoin::address::NetworkUnchecked>> {
        let address = self.client.get_new_address(None, None)?;
        Ok(address)
    }

    /// List unspent transaction outputs (UTXOs)
    pub fn list_utxos(&self) -> Result<Vec<Utxo>> {
        let unspent = self.client.list_unspent(
            Some(self.config.min_confirmations as usize),
            None,
            None,
            None,
            None,
        )?;

        let utxos = unspent
            .into_iter()
            .map(|u| Utxo {
                txid: u.txid,
                vout: u.vout,
                amount: u.amount.to_sat(),
                script_pubkey: u.script_pub_key,
                confirmations: u.confirmations,
            })
            .collect();

        Ok(utxos)
    }

    /// Create a root message (new thread)
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// let txid = wallet.create_root_message("Hello, ANCHOR!")?;
    /// println!("Created message: {}", txid);
    /// ```
    pub fn create_root_message(&self, body: &str) -> Result<Txid> {
        self.create_message(AnchorKind::Text, body.as_bytes(), &[])
    }

    /// Create a reply to an existing message
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// let txid = wallet.create_reply(
    ///     "This is a reply!",
    ///     &parent_txid,
    ///     0,
    /// )?;
    /// ```
    pub fn create_reply(&self, body: &str, parent_txid: &Txid, parent_vout: u8) -> Result<Txid> {
        self.create_message(
            AnchorKind::Text,
            body.as_bytes(),
            &[(*parent_txid, parent_vout)],
        )
    }

    /// Create a message with custom kind and multiple anchors
    pub fn create_message(
        &self,
        kind: AnchorKind,
        body: &[u8],
        anchors: &[(Txid, u8)],
    ) -> Result<Txid> {
        // Get UTXOs
        let utxos = self.list_utxos()?;
        if utxos.is_empty() {
            return Err(WalletError::NoUtxos);
        }

        // Get change address
        let change_address = self.get_new_address()?;
        let change_address = change_address.assume_checked();

        // Build transaction
        let mut builder = TransactionBuilder::new()
            .kind(kind)
            .body_bytes(body.to_vec())
            .fee_rate(self.config.fee_rate)
            .change_script(change_address.script_pubkey());

        // Add anchors
        for (txid, vout) in anchors {
            builder = builder.anchor(*txid, *vout);
        }

        // Add inputs (use first UTXO for simplicity)
        // In a real wallet, you'd implement coin selection
        let utxo = &utxos[0];
        builder = builder.input(utxo.txid, utxo.vout, utxo.amount);

        // Build the transaction
        let anchor_tx = builder.build()?;

        // Sign and broadcast
        let txid = self.sign_and_broadcast(&anchor_tx)?;

        Ok(txid)
    }

    /// Build an unsigned ANCHOR transaction
    ///
    /// Use this for custom signing flows (hardware wallets, etc.)
    pub fn build_transaction(
        &self,
        kind: AnchorKind,
        body: &[u8],
        anchors: &[(Txid, u8)],
    ) -> Result<AnchorTransaction> {
        let utxos = self.list_utxos()?;
        if utxos.is_empty() {
            return Err(WalletError::NoUtxos);
        }

        let change_address = self.get_new_address()?;
        let change_address = change_address.assume_checked();

        let mut builder = TransactionBuilder::new()
            .kind(kind)
            .body_bytes(body.to_vec())
            .fee_rate(self.config.fee_rate)
            .change_script(change_address.script_pubkey());

        for (txid, vout) in anchors {
            builder = builder.anchor(*txid, *vout);
        }

        let utxo = &utxos[0];
        builder = builder.input(utxo.txid, utxo.vout, utxo.amount);

        builder.build()
    }

    /// Sign and broadcast a transaction
    pub fn sign_and_broadcast(&self, anchor_tx: &AnchorTransaction) -> Result<Txid> {
        let hex = anchor_tx.to_hex();
        
        // Sign the transaction
        let signed = self.client.sign_raw_transaction_with_wallet(hex.as_str(), None, None)?;
        
        if !signed.complete {
            return Err(WalletError::TransactionBuild(
                "Failed to sign transaction".to_string()
            ));
        }

        // Broadcast - signed.hex is the signed transaction bytes
        let signed_hex = ::hex::encode(&signed.hex);
        let txid = self.client.send_raw_transaction(signed_hex.as_str())?;

        Ok(txid)
    }

    /// Broadcast a raw transaction hex
    pub fn broadcast(&self, tx_hex: &str) -> Result<Txid> {
        let txid = self.client.send_raw_transaction(tx_hex)?;
        Ok(txid)
    }

    /// Mine blocks (regtest only)
    pub fn mine_blocks(&self, count: u32) -> Result<Vec<bitcoin::BlockHash>> {
        let address = self.get_new_address()?;
        let address = address.assume_checked();
        let hashes = self.client.generate_to_address(count as u64, &address)?;
        Ok(hashes)
    }

    /// Get transaction details
    pub fn get_transaction(&self, txid: &Txid) -> Result<bitcoincore_rpc::json::GetTransactionResult> {
        let tx = self.client.get_transaction(txid, None)?;
        Ok(tx)
    }

    /// Get raw transaction
    pub fn get_raw_transaction(&self, txid: &Txid) -> Result<bitcoin::Transaction> {
        let tx = self.client.get_raw_transaction(txid, None)?;
        Ok(tx)
    }

    /// Get blockchain info
    pub fn get_blockchain_info(&self) -> Result<bitcoincore_rpc::json::GetBlockchainInfoResult> {
        let info = self.client.get_blockchain_info()?;
        Ok(info)
    }

    /// Estimate smart fee
    pub fn estimate_fee(&self, blocks: u16) -> Result<Amount> {
        let result = self.client.estimate_smart_fee(blocks, None)?;
        Ok(result.fee_rate.unwrap_or(Amount::from_sat(1000)))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_config_validation() {
        let config = WalletConfig::new("", "user", "pass");
        assert!(config.validate().is_err());

        let config = WalletConfig::new("http://localhost:18443", "user", "pass");
        assert!(config.validate().is_ok());
    }
}

File: libs/js/anchor-sdk/src/wallet.ts
/**
 * ANCHOR Wallet
 *
 * Node.js wallet implementation using Bitcoin Core RPC.
 */

import {
  AnchorKind,
  AnchorError,
  AnchorErrorCode,
  CarrierType,
  type WalletConfig as WalletConfigType,
  type Balance,
  type Utxo,
  type TransactionResult,
} from './types.js';
import { buildTransaction, getNetwork, type TransactionBuilderOptions } from './transaction.js';
import * as bitcoin from 'bitcoinjs-lib';

/**
 * Bitcoin Core RPC client
 */
class BitcoinRpc {
  private url: string;
  private auth: string;

  constructor(config: WalletConfigType) {
    const walletPath = config.walletName ? `/wallet/${config.walletName}` : '';
    this.url = `${config.rpcUrl}${walletPath}`;
    this.auth = Buffer.from(`${config.rpcUser}:${config.rpcPassword}`).toString('base64');
  }

  async call<T>(method: string, params: unknown[] = []): Promise<T> {
    const response = await fetch(this.url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Basic ${this.auth}`,
      },
      body: JSON.stringify({
        jsonrpc: '1.0',
        id: Date.now(),
        method,
        params,
      }),
    });

    if (!response.ok) {
      throw new AnchorError(
        AnchorErrorCode.RpcError,
        `RPC error: ${response.status} ${response.statusText}`
      );
    }

    const data = (await response.json()) as { result: T; error: unknown };

    if (data.error) {
      throw new AnchorError(AnchorErrorCode.RpcError, `RPC error: ${JSON.stringify(data.error)}`);
    }

    return data.result;
  }
}

/**
 * ANCHOR Wallet for creating and broadcasting messages
 */
export class AnchorWallet {
  private rpc: BitcoinRpc;
  private config: WalletConfigType;
  private bitcoinNetwork: bitcoin.Network;

  constructor(config: WalletConfigType) {
    this.config = {
      feeRate: 1,
      ...config,
    };
    this.rpc = new BitcoinRpc(config);
    this.bitcoinNetwork = getNetwork(config.network);
  }

  /**
   * Get wallet balance
   */
  async getBalance(): Promise<Balance> {
    interface BalanceResult {
      mine: {
        trusted: number;
        untrusted_pending: number;
      };
    }

    const result = await this.rpc.call<BalanceResult>('getbalances');

    const confirmed = Math.round(result.mine.trusted * 1e8);
    const unconfirmed = Math.round(result.mine.untrusted_pending * 1e8);

    return {
      confirmed,
      unconfirmed,
      total: confirmed + unconfirmed,
    };
  }

  /**
   * Get a new receiving address
   */
  async getNewAddress(): Promise<string> {
    return this.rpc.call<string>('getnewaddress');
  }

  /**
   * List unspent outputs (UTXOs)
   */
  async listUtxos(minConfirmations: number = 1): Promise<Utxo[]> {
    interface UnspentResult {
      txid: string;
      vout: number;
      address: string;
      amount: number;
      confirmations: number;
      scriptPubKey: string;
    }

    const result = await this.rpc.call<UnspentResult[]>('listunspent', [minConfirmations]);

    return result.map((u) => ({
      txid: u.txid,
      vout: u.vout,
      value: Math.round(u.amount * 1e8),
      scriptPubKey: u.scriptPubKey,
    }));
  }

  /**
   * Create a root message (new thread)
   */
  async createRootMessage(text: string, carrier?: CarrierType): Promise<TransactionResult> {
    return this.createMessage({
      kind: AnchorKind.Text,
      body: text,
      carrier,
    });
  }

  /**
   * Create a reply to an existing message
   */
  async createReply(
    text: string,
    parentTxid: string,
    parentVout: number = 0,
    carrier?: CarrierType
  ): Promise<TransactionResult> {
    return this.createMessage({
      kind: AnchorKind.Text,
      body: text,
      anchors: [{ txid: parentTxid, vout: parentVout }],
      carrier,
    });
  }

  /**
   * Create a permanent root message using Stamps carrier
   */
  async createPermanentMessage(text: string): Promise<TransactionResult> {
    return this.createRootMessage(text, CarrierType.Stamps);
  }

  /**
   * Create a permanent reply using Stamps carrier
   */
  async createPermanentReply(
    text: string,
    parentTxid: string,
    parentVout: number = 0
  ): Promise<TransactionResult> {
    return this.createReply(text, parentTxid, parentVout, CarrierType.Stamps);
  }

  /**
   * Create a message with custom options
   */
  async createMessage(
    options: TransactionBuilderOptions['message'] & { carrier?: CarrierType }
  ): Promise<TransactionResult> {
    // Get UTXOs
    const utxos = await this.listUtxos(0);
    if (utxos.length === 0) {
      throw new AnchorError(AnchorErrorCode.NoUtxos, 'No UTXOs available');
    }

    // Get change address
    const changeAddress = await this.getNewAddress();

    // For Stamps, we need more UTXOs
    const inputCount = options.carrier === CarrierType.Stamps ? Math.min(2, utxos.length) : 1;
    const inputs = utxos.slice(0, inputCount);

    // Build transaction
    const { psbt, anchorVout } = buildTransaction({
      network: this.config.network,
      feeRate: this.config.feeRate,
      inputs,
      changeAddress,
      message: options,
    });

    // Sign and broadcast
    const result = await this.signAndBroadcast(psbt);

    // Determine carrier used
    const carrier = options.carrier ?? CarrierType.OpReturn;

    return {
      ...result,
      vout: anchorVout,
      carrier,
    };
  }

  /**
   * Sign a PSBT using Bitcoin Core wallet
   */
  async signPsbt(psbt: bitcoin.Psbt): Promise<bitcoin.Psbt> {
    const hex = psbt.toHex();

    interface SignResult {
      psbt: string;
      complete: boolean;
    }

    const result = await this.rpc.call<SignResult>('walletprocesspsbt', [hex]);

    if (!result.complete) {
      throw new AnchorError(AnchorErrorCode.SigningError, 'Failed to sign transaction');
    }

    return bitcoin.Psbt.fromBase64(result.psbt, {
      network: this.bitcoinNetwork,
    });
  }

  /**
   * Sign and broadcast a transaction
   */
  async signAndBroadcast(psbt: bitcoin.Psbt): Promise<{ txid: string; hex: string }> {
    // Sign
    const signedPsbt = await this.signPsbt(psbt);

    // Finalize
    signedPsbt.finalizeAllInputs();

    // Extract transaction
    const tx = signedPsbt.extractTransaction();
    const hex = tx.toHex();

    // Broadcast
    const txid = await this.rpc.call<string>('sendrawtransaction', [hex]);

    return { txid, hex };
  }

  /**
   * Broadcast a raw transaction
   */
  async broadcast(hex: string): Promise<string> {
    return this.rpc.call<string>('sendrawtransaction', [hex]);
  }

  /**
   * Mine blocks (regtest only)
   */
  async mineBlocks(count: number = 1): Promise<string[]> {
    const address = await this.getNewAddress();
    return this.rpc.call<string[]>('generatetoaddress', [count, address]);
  }

  /**
   * Get blockchain info
   */
  async getBlockchainInfo(): Promise<{
    chain: string;
    blocks: number;
    headers: number;
    bestblockhash: string;
  }> {
    return this.rpc.call('getblockchaininfo');
  }

  /**
   * Get raw transaction
   */
  async getRawTransaction(txid: string, verbose: boolean = true): Promise<unknown> {
    return this.rpc.call('getrawtransaction', [txid, verbose]);
  }
}

/**
 * Create a wallet instance
 */
export function createWallet(config: WalletConfigType): AnchorWallet {
  return new AnchorWallet(config);
}

/**
 * Create wallet config helpers
 */
export const WalletConfigHelper = {
  mainnet(rpcUrl: string, rpcUser: string, rpcPassword: string): WalletConfigType {
    return { rpcUrl, rpcUser, rpcPassword, network: 'mainnet' };
  },
  testnet(rpcUrl: string, rpcUser: string, rpcPassword: string): WalletConfigType {
    return { rpcUrl, rpcUser, rpcPassword, network: 'testnet' };
  },
  signet(rpcUrl: string, rpcUser: string, rpcPassword: string): WalletConfigType {
    return { rpcUrl, rpcUser, rpcPassword, network: 'signet' };
  },
  regtest(rpcUrl: string, rpcUser: string, rpcPassword: string): WalletConfigType {
    return { rpcUrl, rpcUser, rpcPassword, network: 'regtest' };
  },
};

// Re-export as WalletConfig for convenience
export { WalletConfigHelper as WalletConfig };
